examples/webgl/src/main.rs

```diff
@@ -45,45 +34,37 @@ impl Component for App {
         }
     }
 
-    fn rendered(&mut self, ctx: &Context<Self>, first_render: bool) {
+    fn rendered(&mut self, ctx: &Context<Self>, _first_render: bool) {
         // Once rendered, store references for the canvas and GL context. These can be used for
         // resizing the rendering area when the window or canvas element are resized, as well as
         // for making GL calls.
-
         let canvas = self.node_ref.cast::<HtmlCanvasElement>().unwrap();
-
         let gl: GL = canvas
             .get_context("webgl")
             .unwrap()
             .unwrap()
             .dyn_into()
             .unwrap();
-
-        self.gl = Some(gl);
-
-        // In a more complex use-case, there will be additional WebGL initialization that should be
-        // done here, such as enabling or disabling depth testing, depth functions, face
-        // culling etc.
-
-        if first_render {
-            // The callback to request animation frame is passed a time value which can be used for
-            // rendering motion independent of the framerate which may vary.
-            let handle = {
-                let link = ctx.link().clone();
-                request_animation_frame(move |time| link.send_message(Msg::Render(time)))
-            };
-
-            // A reference to the handle must be stored, otherwise it is dropped and the render
-            // won't occur.
-            self._render_loop = Some(handle);
-        }
+        self.gl = Some(Rc::new(gl));
+        self.render_gl(ctx.link());
     }
 }
 
 impl App {
-    fn render_gl(&mut self, timestamp: f64, link: &Scope<Self>) {
+
+    fn request_animation_frame(f: &Closure<dyn FnMut()>) {
+        window().unwrap()
+            .request_animation_frame(f.as_ref().unchecked_ref())
+            .expect("should register `requestAnimationFrame` OK");
+    }
+
+    fn render_gl(&mut self, _link: &Scope<Self>) {
+        log!("This should log only once -- not once per frame");
+        
         let gl = self.gl.as_ref().expect("GL Context not initialized!");
```

Comment:
With the change, you should be able to pass `gl` as a normal function parameter, instead of akwardly passing (and unwrapping) it through self. Also feel free to just remove `link` from the function parameters.

Reply:
I've been contemplating the issue of how much game state the Yew application needs access to.  This is inherited, but it's worth highlighting.  Thank you.

I'm guessing the whole gl process could go into an Agent/WebWorker/thread, along with the WebSocket/WebRTC client as expect high-traffic between the game and communications IO.  Very likely scaling would want to take advantage of available threads.  That would require Arc and atomics or Mutex.

Reply:
The Yew app thread would be left alone with the DOM.